File: skel/lab01/java/task-2/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class Main {
    static class Task {
        public final static String INPUT_FILE = "in";
        public final static String OUTPUT_FILE = "out";

        double n;

        public void solve() {
            readInput();
            writeOutput(getResult());
        }

        private void readInput() {
            try {
                Scanner sc = new Scanner(new File(INPUT_FILE));
                n = sc.nextDouble();
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(double x) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%.4f\n", x);
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private double getResult() {
            // TODO: Calculeaza sqrt(n) cu o precizie de 10^-3.
            // Precizie de 10^(-x) inseamna |valoarea_ta - valoarea_reala| < 10^(-x).

            double left = 1;
            double right = n;
            
            if(n < 1) {
                left = n;
                right = 1;
            }
            double mid = -1;

            System.out.println("n = " + n);
            System.out.println("right = " + right);

            while(left <= right) {
                mid = (left + right) / 2;
                System.out.println("mid = " + mid);

                if(mid * mid - n > 0.0001) {
                    right = mid;
                } else if(mid * mid - n < -0.0001) {
                    left = mid;
                } else{
                    break;
                }

                
            }


            return mid;
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}

File: skel/lab01/cpp/task-2/main.cpp
#include <bits/stdc++.h>
using namespace std;

class Task {
public:
    void solve() {
        read_input();
        print_output(get_result());
    }

private:
    double n;

    void read_input() {
        ifstream fin("in");
        fin >> n;
        fin.close();
    }

    double get_result() {
        // TODO: Calculati  sqrt(n) cu o precizie de 0.001
        // Precizie de 10^-x = |rezultatul vostru - rezultatul corect| <= 10^-x
        printf("n = %f\n", n);
        double left, right;
        if (n >= 1) {
            left = 0;
            right = n;
        } else {
            left = n;
            right = 1;
        }
        while (1) {
            double mid = (left + right) / 2;
            printf("%f\n", mid);
            // double result = n >= 1 ?: n - mid * mid;
            if (mid * mid - n > 0.0001) {
                right = mid;
            } else if (mid * mid - n < -0.0001) {
                left = mid;
            } else {
                return mid;
            }
        }
    }

    void print_output(double result) {
        ofstream fout("out");
        fout << setprecision(4) << fixed << result;
        fout.close();
    }
};

// [ATENTIE] NU modifica functia main!
int main() {
    // * se aloca un obiect Task pe heap
    // (se presupune ca e prea mare pentru a fi alocat pe stiva)
    // * se apeleaza metoda solve()
    // (citire, rezolvare, printare)
    // * se distruge obiectul si se elibereaza memoria
    auto* task = new (nothrow) Task(); // hint: cppreference/nothrow
    if (!task) {
        cerr << "new failed: WTF are you doing? Throw your PC!\n";
        return -1;
    }
    task->solve();
    delete task;
    return 0;
}
